c7n/resources/ec2.py

```diff
@@ -920,6 +920,77 @@ def process_resource_set(self, client, resources):
             r[self.annotation] = info_map.get(r['InstanceId'], {})
 
 
+@EC2.filter_registry.register('ssm-compliance')
+class SsmCompliance(Filter):
+    """Filter ec2 instances by their ssm compliance status.
+
+    :Example:
+
+    Find non-compliant ec2 instances.
+
+    .. code-block:: yaml
+
+        policies:
+          - name: ec2-ssm-compliance
+            resource: ec2
+            filters:
+              - type: ssm-compliance
+                compliance_types:
+                  - Association
+                  - Patch
+                states:
+                  - NON_COMPLIANT
+    """
+    schema = type_schema(
+        'ssm-compliance',
+        **{'required': ['compliance_types'],
+           'compliance_types': {'type': 'array', 'items': {'type': 'string'}},
```

Comment:
FYI, leaving compliance_type's open-ended to support custom types.  [AWS documentation](https://docs.aws.amazon.com/systems-manager/latest/userguide/sysman-compliance-about.html) states you can have up to 10 custom compliance types associated with an instance, e.g. `Custom:MyOrgCheck`